first pass at procedures (#105)
kwilteam/docs: docs/kuneiform/procedures.mdx

---
sidebar_position: 3
sidebar_label: "Procedures"
id: kuneiform-procedures
title: Kuneiform Procedures
description: An outline of Kuneiform procedures.
slug: /kuneiform/procedures
---

Kuneiform supports a basic procedural language for building logic into schemas. Procedures can be used for building access
control logic, performing arithmetic, and managing control flow.

## Declaring Procedures

Procedures are declared using the `procedure` keyword. They can be given a list of parameters and their corresponding types,

> return types, and access modifiers.

Reviewer: Add a note that if a procedure returns data, the return must be declared in the signature?

Author: I feel like that is sort've implied...

I'll add a note about strong typing

// be called without submitting a blockchain transaction
procedure get_user($id uuid) public view returns (name text, age int) {
    // procedure implementation
}
```

Procedures can also return tables, which are a collection of many records. The below example returns all user IDs and names
of a certain age:

```ts
// get_users takes one integerm and will return a table with two columns:
// "name" of type text, and "age" of type int
// It has a view tag, which means it does not mutate state, and can
// be called without submitting a blockchain transaction
procedure get_users($age int) public view returns table(name text, age int) {
    // procedure implementation
}
```

If a procedure does not return anything, the return should not be specified:

```ts
// create_user takes a text and int, and returns nothing.
// It does not have a view tag so that it can mutate state,
// and therefore can only be called with a blockchain transaction.
procedure create_user($name text, $age int) public {
    // procedure implementation
}
```

## Access Modifiers

Procedures can be given access modifiers to specify how/when they can be called. All procedures must be specified

> as `public` or `private`: `public` procedures can be called externally, while `private` procedures can only be called

Reviewer: I think it is easier to read if the access modifiers are presented in a table. The way this reads, it is hard to follow all the supported tags.

Author: I sort've disagree, because a table conveys that they are all options. For example, in the table, since `view` and `owner` are there, it makes me think that I can use those instead of public. We even saw fractal trip up on this. I think this more clearly conveys that `owner` and `view` are not replacements for public and private

**Arrays use 1-based indexing.**

```ts
procedure handle_array() public view returns (val text) {
    $arr text[] := ['hello', 'world'];
    $arr := array_append($arr, 'goodbye');

    $len int := array_length($arr); // $len == 3

    // return the first value.
    return $arr[1];
}
```

### If/Then

```ts
procedure age_conditional($age int) public {
    if $age < 0 {
        // do something
    } elseif $age < 18 {
        // do something
    } else {
        // something else
    }
}
```

### For Loop

For loops can be declared over 3 different targets: arrays, ranges, and SQL statements.

#### Array Loop

```ts
procedure iterate_over_array($names text[]) public {
    for $name in $names {
        // $name is of type `text`
    }
}
```

#### Range Loop

```ts
procedure iterate_over_range() public {
    // iterating over all numbers from 1 to 25
    for $i in 1..25 {
        // $i is type int
    }
}

procedure iterate_range_dynamic($start int, $end int) public {
    // can also iterate over dynamic ranges
    for $i in $start..$end {
        // ...
    }
}
```

#### SQL Loops

> You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.

Reviewer: This seems more precise:

```suggestion
You can also iterate over each record returned from a SQL query. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
```

```ts
procedure iterate_over_sql() public {
    for $row in SELECT id, name, age FROM users {
        // can access $row.id, $row.name, and $row.age
    }
}
```

Procedures that return tables can also be iterated over:

```ts
procedure get_users() public view returns table(id uuid, name text, age int) {
    // ...
}

procedure iterate_over_users() public {
    for $row in SELECT * FROM get_users() {
        // can access $row.id, $row.name, and $row.age
    }
}
```

#### Breaking From a Loop

To conditionally exit a loop, you can use `break`:

```ts
procedure get_users() public {
    $count int := 0;
    for $row in SELECT * FROM users {
        // do something

        // break if too many records have been read
        $count := $count + 1;
        if $count >= 50 {

>             break

Reviewer: Does this `break` need a semicolon? It feels weird that every other line in this language ends with a semicolon, but `break` does not?
